### DISCUSSION TITLE
Error detection for valid json that doesn't match the c++ types

### DISCUSSION THREAD

**Discussion by madeso** [nlohmann/json#2883] [Q&A] [answered]
Hello,

I'm parsing json files that are written by hand and here errors are somewhat common so I need good error handling.
Is there a way to get detailed information about where in the file/string a certain node was specified so I can generate a better error message?

Example json of a list of name/ages:
```json
[
    {"naem": "eve", "aeg": 16},
    {"name": "adam", "age": 15},
    {},
    {"name": 42, "age": "john"},
    {"name": "maria", "age": 9}
]
```

Executable example
```cpp
#include <string>
#include <string_view>
#include <vector>
#include <iostream>

#include <nlohmann/json.hpp>
using json = nlohmann::json;

////////////////////////// types

struct Person
{
    std::string name; double age;
    Person() : name(""), age(0.0) {}
    Person(const std::string& n, double a) : name(n), age(a) {}
};

void print(const std::vector<Person>& persons)
{
    for(const auto& p: persons)
    {
        std::cout << p.name << " " << p.age << "\n";
    }
}

////////////////////////// parsing functions

// for simplicity, log errors and continue
std::vector<Person> read_persons(const std::string_view& source)
{
    auto r = std::vector<Person>{};

    auto root = json::parse(source);
    if(root.is_array() == false)
    {
        std::cout << "root element is not an array\n";
        return r;
    }

    int next_index = 0;
    for(auto& p: root)
    {
        next_index += 1;
        const auto index = next_index - 1;
        const auto name = p.find("name");
        const auto age = p.find("age");

        if(name == p.end())
        {
            std::cout << "element at " << index << " is missing property name\n";
            continue;
        }

        if(name->is_string() == false)
        {
            std::cout << "element at " << index << " has a non string name\n";
            continue;
        }

        const auto name_value = name->get<std::string>();

        if(age == p.end())
        {
            std::cout << "element at " << index << " is missing property age\n";
            continue;
        }

        if(age->is_number() == false)
        {
            std::cout << "element at " << index << " has age that is not a number\n";
            continue;
        }

        const auto age_value = age->get<double>();

        r.emplace_back(name_value, age_value);
    }

    return r;
}

////////////////////////// Sample

constexpr std::string_view persons_json =
R"(
    [
        {"naem": "eve", "aeg": 16},
        {"name": "adam", "age": 15},
        {},
        {"name": 42, "age": "john"},
        {"name": "maria", "age": 9}
    ]
)";

int main()
{
    const auto persons = read_persons(persons_json);
    print(persons);

    return persons.size();
}
```
[godbolt](https://godbolt.org/z/451T6PfKM)

I don't care much about the verbosity of the parsing code as it is generated from metadata.

**Comment by nlohmann** [comment, marked as answer]
I think you are less interested in diagnostics of the parser, but rather in the assumed structure of the generated JSON values. For this:

- You can use the latest (yet unreleased) version from the `develop` branch and define `JSON_DIAGNOSTICS` to `1` to get better diagnostics. See https://json.nlohmann.me/home/exceptions/#extended-diagnostic-messages for more information.
- You could also simplify your code by providing a `from_json` function for struct `Person`, see https://json.nlohmann.me/features/arbitrary_types/

**Reply by madeso** [reply]
I have a `from_json` in the example but removed it for brevity from the post as it didn't add anything. Setting diagnostic to 1 improves the output a bit but not much (I assume the godbolt version trunk is the same as develop).
```[json.exception.out_of_range.403] (/0) key 'name' not found```

Don't get me wrong, this error is nice, but comparing to my custom code where I run a edit-distance search for all unused keys on the object and display suggestions and display a "readable" path to the item, usually by adding id or name keys to the the arrays indices so I get `[4(id: "dog")][2](id: "cat") is missing name, but has invalid key 'naem'` instead of `[4][2] key 'name' not found`, using the build in parser is unfortunately not an improvement.

I would like to add line + byte offset to this error message finding and correcting the item at `[50][4]` is a more cumbersome than it could be.

**Reply by nlohmann** [reply]
The JSON Pointer (`/0`) gives you a hint that the error affects the first element of an array. I understand that this may be insufficient for many use cases, but at the same time, the library can only implement so much application logics. If your code wants to access a non-existing key `name`, then it's up to you to decide what this means.

The library has no link to the input after parsing - the values do not "know" from which line or column they have been parsed.

**Reply by madeso** [reply]
Is there a way some way to get access to the input after parsing, by perhaps using the Sax parser, a callback or a custom string/bool/etc type? From what I can tell by reading the documentation there is none, but perhaps I am missing something.

If it isn't possible, would a pull request, for example, that adds it be accepted or it it outside the scope of `nlohmann::json`?

**Reply by nlohmann** [reply]
I am mostly concerned about adding such a feature without breaking the API.

**Reply by gregmarr** [reply]
I could see adding two integers as part of the `JSON_DIAGNOSTICS` option for line number and column assuming that information is available in the parser and the parser could set those when it creates the object.  Then that info could be added to the diagnostics output if it's non-zero.  That might provide a bit more information for the user, or the program using the library could access that information and provide the more advanced type of reporting.

**Reply by nlohmann** [reply]
Right - since the extended diagnostics have not yet been released, we can use this macro to separate these changes from the existing public API. I am, however, a bit reluctant to really store the source location in a `basic_json` object, because already the parent relation needed touching a lot of places, and we had to fix some bugs afterwards.

**Reply by gregmarr** [reply]
The source location should be a lot less fragile as it's not a pointer to another object.  For example, you don't have to reset it in all the children of an array when the array resizes.  It's probably not trivial, but not as bad as the parent.  It would also only set by `parse()`, not during arbitrary creation.

**Reply by madeso** [reply]
I'm not familiar with the inner structure and what exactly `basic_json` represents, but I thought it would be enough to store it on the actual [values types](https://json.nlohmann.me/features/types/). To get the location for a certain type you would wrap that type in a `with_location<T>`. Parsing needs to send the location if it's needed but if constexpr is available there wouldn't be any additional branching in the parsing.

Usage would differ if the end user choose to include the line numbers but that might be unavoidable.

**Reply by nlohmann** [reply]
I am afraid touching the value types could even be more work. But I’m happy to be proven differently.

**Reply by gregmarr** [reply]
The values type is a union so I don't think that would be an appropriate place.

**Reply by madeso** [reply]
Hello... I looked into replacing the value types with a `with_location<T>` type and I pretty soon hit a dead end since at least the number type needs to be both trivially constructible and constructible from a double and I can't seem to figure out how to write the type without modifying the usage.

I also realized that if I switch to xml, both editor support for xsd and xpath seems to be better than the json equivalent so I won't need this change for the foreseeable future. Thank you both for the shown interest and help.

**Comment by marton78** [comment]
One additional, thing, @madeso: Don't pass `string_view` by const ref, pass it by value! That is, change the following line:

```c++
std::vector<Person> read_persons(std::string_view source)
```
